Add C++ header generation

MoonlightSentinel opened the pull request:
Squashed rebase of #9971 without the changes to the existing C++ tests (as proposed in https://github.com/dlang/dmd/pull/9971#issuecomment-586764121).

CC @edi33416, @TurkeyMan, @wilzbach

ibuclaw committed:
Generate C++ header files from public extern (C++) declarations

The -HC switch let's dmd output header files from existing extern(C++)
files to ease C++ interoperation. Use -HCd=<directory> / -HCf=<file>
to select the directory/file where the header should be generated.

The following commits were squashed from #9971 without the changes
to the existing C++ interface/test suite:

Testing D to H tool

Remove unused and redundant dmd headers

Improve dtoh

Factor out compiler specific parts
Attempt to make cross platform

Fix stuff

Add dtoh as a compiler switch

Update code after rebase

Compile frontend srcs and write to flie

Add dtoh enum test

Add free functions dtoh test

Add unittest block dtoh test

Add VarDeclaration dtoh test

Add AliasDeclaration dtoh test

Add AnonDeclaration dtoh test

Add StructDeclaration dtoh test

Add ClassDeclaration dtoh test

Add TemplateDeclaration dtoh test

Add debug prints

Fix struct field void initializer

Add ignoredModule tmp helper

Add tmp workaround for alias forward decl and >> template issue

Add support for enum base type

Add support for alignment of pow2

Fix alignment issues

Generate enums based on -extern-std version

Update tool auto-gen message

Update tests

Code cleanup

CPP does not support covariant const methods

Correctly write ctor declarations

Generate default args for literals

Generate dtor and copy ctor declarations

Rebase with master

Add compiler switch to cli.d

Update tests

Add support for @disable

Improve default ctor generation

Update struct decl test

Use member initializer list for default ctor definition

Co-Authored-By: Eduard Staniloiu <edi33416@gmail.com>

(Rebase + Fixup)
Co-Authored-By: MoonlightSentinel <moonlightsentinel@disroot.org>

MoonlightSentinel committed:
Address style feedback by thewilsonator

thewilsonator commented on src/dmd/dtoh.d:
```diff
@@ -0,0 +1,1724 @@
+        __gshared static Identifier ID; ////Identifier
+        __gshared static Identifier Visitor;
+        __gshared static Identifier ParseTimeVisitor;
+        __gshared static void _init()
```
`__gshared` on a function doesn't make sense. Just remove it.

thewilsonator commented on src/dmd/dtoh.d:
```diff
@@ -0,0 +1,1724 @@
+            scope(exit) printf("[AST.AttribDeclaration exit] %s\n", pd.toChars());
+        }
+        Dsymbols* decl = pd.include(null);
+        if (decl)
```
```
if (!decl)
    return;
...
```

thewilsonator commented on src/dmd/dtoh.d:
```diff
@@ -0,0 +1,1724 @@
+        buf.writestring(sd.isUnionDeclaration() ? "union" : "struct");
+        pushAlignToBuffer(sd.alignment);
+        buf.writestring(sd.ident.toChars());
+        if (sd.members)
```
```
if (!sd.members)
{
    buf.writestring(";\n\n");
    return;
}
...
```

thewilsonator commented on src/dmd/dtoh.d:
```diff
@@ -0,0 +1,1724 @@
+
+    private void popAlignToBuffer(uint alignment)
+    {
+        if (alignment != uint.max)
```
```
if (alignment == uint.max)
    return;
...
```

thewilsonator replied:
This was marked resolved but not addressed.

MoonlightSentinel replied:
Fixed

thewilsonator commented on src/dmd/dtoh.d:
```diff
@@ -0,0 +1,1724 @@
+            printf("[includeSymbol(AST.Dsymbol) enter] %s\n", ds.toChars());
+            scope(exit) printf("[includeSymbol(AST.Dsymbol) exit] %s\n", ds.toChars());
+        }
+        if (cast(void*)ds !in visited)
```
```
if (cast(void*)ds in visited)
    return;
...
```

thewilsonator commented on src/dmd/dtoh.d:
```diff
@@ -0,0 +1,1724 @@
+
+            includeSymbol(cd.baseClass);
+        }
+        if (cd.members)
```
if (!cd.members)
{
    buf.writestring(";\n\n");
    return;    
}

thewilsonator commented on src/dmd/dtoh.d:
```diff
@@ -0,0 +1,1724 @@
+        }
+        else
+        {
+            if (hasBaseType)
```
```
else if (hasBaseType)
{
    ...
}
else
{
      buf.writestring("enum ");
      buf.writestring(ident);
}
```

thewilsonator commented on src/dmd/dtoh.d:
```diff
@@ -0,0 +1,1724 @@
+                buf.writestring(ident);
+            }
+        }
+        if (ed.members)
```
```
if (!ed.members)
{
     buf.writestring(";\n\n");
    return;
}
...
```

thewilsonator commented on src/dmd/dtoh.d:
```diff
@@ -0,0 +1,1724 @@
+            printf("[enumToBuffer(AST.EnumDeclaration) enter] %s\n", ed.toChars());
+            scope(exit) printf("[enumToBuffer(AST.EnumDeclaration) exit] %s\n", ed.toChars());
+        }
+        if (ed.isSpecial())
```
```
if (!ed.isSpecial())
{
    buf.writestring(ed.toChars());
    return;
}
...
```

thewilsonator reviewed (commented):
Thanks! First review pass, Indentation:

thewilsonator commented on src/dmd/dtoh.d:
```diff
@@ -0,0 +1,1724 @@
+    __gshared static Identifier c_longlong;
+    __gshared static Identifier c_ulonglong;
+    __gshared static Identifier c_long_double;
+    //version(BUILD_COMPILER)
```
Its probably best to remove all the commented out version(BUILD_COMPILER) and put it behind a (hidden) compiler switch instead.

MoonlightSentinel replied:
AFAICT the purpose of this spearate program was to only generate header files without actual codegen?

Assuming this is the case one could achieve the same behaviour using `-c -o-`(?)

thewilsonator replied:
No the purpose of that program was to do all the weird stuff that the DMD headers need see `isVisitorClass`, `isIgnoredModule` and friends. And separate it from the behaviour that compiler users would expect from the `-HC` switch.

MoonlightSentinel replied:
Thanks. I've hidden this functionality behind an enum `isBuildingCompiler` until we define an appropriate name for this  flag. 

But the following code snippet got me thinking whether special casing for DMD is a good idea or if there's a deeper problem with the current header generation.

https://github.com/dlang/dmd/blob/3585ac12f887d90c7f621c1b65d7af72a46bd8db/src/dmd/dtoh.d#L1294-L1329

thewilsonator commented on src/dmd/dtoh.d:
```diff
@@ -0,0 +1,1724 @@
+    #endif
+    `);
+
+    debug(HEADER) buf.write(&check);
```
Why is this not using Debug_DtoH?

MoonlightSentinel replied:
Because i missed that identifier.

thewilsonator commented on src/dmd/dtoh.d:
```diff
@@ -0,0 +1,1724 @@
+
+    OutBuffer check;
+    check.writestring(`
+    #if OFFSETS
```
Please put all if this in the debug clock below.

MoonlightSentinel replied:
`check` is used unconditionally below

thewilsonator replied:
I meant all the uses of it, everything between line 230-269.

MoonlightSentinel replied:
That change results in wrong headers for non-debug builds....

Geod24 commented on src/dmd/root/dcompat.h:
```diff
@@ -24,6 +24,19 @@ struct DArray
 };
 
+template<>
+struct DArray<char>
```
Why isn't this using `DString` ?

MoonlightSentinel replied:
Probably because these changes predate `DString`

MoonlightSentinel replied:
On a second look, DString/DArray is missing some of the constructors / operator overloads. Should i add them?

Geod24 replied:
Sure. Which ones are missing ?

MoonlightSentinel replied:
```
DArray<char>& operator=(const char str[]) { ptr = str; length = strlen(str); }
```

Ignore the commant regarding the constructor, I forgot that C++ does not differentiate between `const char str[]` and `const char* str`

MoonlightSentinel replied:
Wait, even that operator overload is not required because of C++ implicit conversion ... feature. Sorry for the noise

thewilsonator commented on src/dmd/dtoh.d:
```diff
@@ -0,0 +1,1724 @@
+            else
+            {
+                buf.writestring(" /* const */");
+            }
```
you can drop the braces here.

Geod24 commented on test/compilable/extra-files/dtoh-postscript.sh:
```diff
@@ -0,0 +1,5 @@
+#!/usr/bin/env bash
+
+diff -pu --strip-trailing-cr ${EXTRA_FILES}/${TEST_NAME}.out ${OUTPUT_BASE}.out
```
Use `.h` so we get syntax highlighting

Geod24 commented on test/compilable/extra-files/dtoh-postscript.sh:
```diff
@@ -0,0 +1,5 @@
+
+diff -pu --strip-trailing-cr ${EXTRA_FILES}/${TEST_NAME}.out ${OUTPUT_BASE}.out
+
+rm -f ${OUTPUT_BASE}.out
```
Use `trap`

MoonlightSentinel replied:
Maybe this should be omitted entirely, see https://github.com/dlang/dmd/pull/9971#discussion_r304808436

MoonlightSentinel replied:
I think these tests should use `TEST_OUTPUT` anyway s.t. we could use #10781 to handle platform based differences (and have fewer bash tests)

MoonlightSentinel replied:
Done

Geod24 commented on src/dmd/globals.d:
```diff
@@ -239,6 +239,10 @@ extern (C++) struct Param
 
+    bool doCxxHdrGeneration;            // write 'Cxx header' file
+    const(char)[] cxxhdrdir;            // write 'header' file to docdir directory
+    const(char)[] cxxhdrname;           // write 'header' file to docname
```
Update to header files seems to be missing ?

Geod24 commented on src/dmd/dtoh.d:
```diff
@@ -0,0 +1,1724 @@
+/**
+ * Compiler implementation of the
+ * $(LINK2 http://www.dlang.org, D programming language).
```
Please provide a more useful description

thewilsonator commented on src/dmd/dtoh.d:
```diff
@@ -0,0 +1,1724 @@
+                buf.printf(" {}\n");
+            }
+
+            version (none)
```
any ideas what this version none ins for?

MoonlightSentinel replied:
Not sure but the versioned code generates an additional constructor to set all fields, e.g.

```
S2(_d_int a, _d_int b, _d_long c) { this->a = a; this->b = b; this->c = c; }
```

for 

```
extern (C++) struct S2
{
    int a = 42;
    int b;
    long c;

    this(int a) {}
}
```

Geod24 commented on src/dmd/dtoh.d:
```diff
@@ -0,0 +1,1724 @@
+ * Compiler implementation of the
+ * $(LINK2 http://www.dlang.org, D programming language).
+ *
+ * Copyright:   Copyright (C) 1999-2019 by The D Language Foundation, All Rights Reserved
```
```suggestion
 * Copyright:   Copyright (C) 1999-2020 by The D Language Foundation, All Rights Reserved
```

Geod24 commented on src/dmd/dtoh.d:
```diff
@@ -0,0 +1,1724 @@
+
+private struct DMDType
+{
+    __gshared static Identifier c_long;
```
`__gshared` and `static` are redundant, drop the `static`

Geod24 commented on src/dmd/dtoh.d:
```diff
@@ -0,0 +1,1724 @@
+    __gshared static Identifier c_ulonglong;
+    __gshared static Identifier c_long_double;
+    //version(BUILD_COMPILER)
+    //{
```
?

thewilsonator replied:
Thats a leftover refactor of mine from Iain's PR to make it a separate program and remove the DMD specific bits of the diff. https://github.com/dlang/dmd/pull/10801#discussion_r380643827

Geod24 commented on src/dmd/dtoh.d:
```diff
@@ -0,0 +1,1724 @@
+    OutBuffer buf;
+    buf.writestring("#pragma once\n");
+    buf.writeByte('\n');
+    buf.printf("// Automatically generated by dmd -HC\n");
```
I'm wondering whether we should include the DMD version. 
On one hand it's informative, on the other hand it would need to needless diff in the event a generated header is checked in and its generation is part of the process (it doesn't make sense to me, but I'm sure someone will do it).
Same issue would arise with a timestamp. However I can see a value in knowing a bit more about how the header was generated.

Last but not least, having the full command line might be helpful too. Things like `-version` and `-debug` will affect the output, won't they ?

MoonlightSentinel replied:
Indeed, that could be useful to track down errors but also annoying for diffs (and maybe reproducible builds). Maybe this behavior should be customizable via an optional argument to `-HC`. But i think this could be added later.

> Last but not least, having the full command line might be helpful too. Things like `-version` and `-debug` will affect the output, won't they ?

I think so.

ibuclaw replied:
At the very least, use `__VENDOR__`, as it's not only dmd that would use this option.

Geod24 reviewed (commented):
We could use a changelog entry describing the user-facing interface.
Things that immediately come to mind:
- What happens when doing `dmd a.d b.d c.d -HC` ?
- Mention the scope (only applies to `extern(C)` and `extern(C++)` declarations, ignore everything else)
- Mention the limitations (semantic not run)

Geod24 commented:
As a side note, something that would be a life-saver for people using C++ interop easily is an automatic `offsetof` / `sizeof` check (in the form of `static assert`). But currently semantic isn't run so that's not doable.

MoonlightSentinel committed:
Address Geod24's feedback

- Documentation for dtoh.d
- Removed redundant static
- Add missing members to globals.h
- Generate .h instead of .out files in tests
- Add a changelog

MoonlightSentinel commented:
> As a side note, something that would be a life-saver for people using C++ interop easily is an automatic `offsetof` / `sizeof` check (in the form of `static assert`). But currently semantic isn't run so that's not doable.

This could be added later. But for now the main goal is to integrate an initial version s.t. @TurkeyMan  and other people can start contributing as well.

MoonlightSentinel committed:
Replace `version (BUILD_COMPILER)` with a runtime check

MoonlightSentinel commented:
I've added a preliminary changelog but I'm unsure whether we should promote it right away.

Some remaining TODO's:

- Actually implement generating separate header files (the directory passed to `-HDd` is currently treated as a prefix for the file passed to `-HDf` 
- Cleanup generated header files (e.g. decide whether `#define` or `typedef` is used)
- (Opt-in) metadata as proposed in https://github.com/dlang/dmd/pull/10801#discussion_r380651560
- Missing semantic analysis

But these should mostly be resolved in further PR's.

wilzbach commented:
> but I'm unsure whether we should promote it right away.

Maybe state "experimental" support? So that it's clear that not everything is fully tested and bug reports are welcome.

> But these should mostly be resolved in further PR's.

Yes, please keep this one as small/low-effort as possible so that people can finally start testing it. Thank you very much for taking this to the finish line!!

MoonlightSentinel commented:
> Maybe state "experimental" support? So that it's clear that not everything is fully tested and bug reports are welcome.

Done

> Yes, please keep this one as small/low-effort as possible so that people can finally start testing it. Thank you very much for taking this to the finish line!!

Gotcha. Maybe such big features should be developed on a dedicated branch in the future. That would offer similar benefits to a draft PR while also keeping the work publicly accessible (everbody can make a small PR without adopting the entire stalled PR)

RazvanN7 commented:
@Geod24 Semantic is run before generating the cpp header files.

wilzbach commented:
> Maybe such big features should be developed on a dedicated branch in the future. That would offer similar benefits to a draft PR while also keeping the work publicly accessible (everbody can make a small PR without adopting the entire stalled PR)

FYI: We have tried this before and even made nightly builds of these branches available. No one ever downloaded those (I guess it also was a problem of people not knowing about it). Additionally, you run the high risk of the branch having lots of merge conflicts with master, so IMHO the best thing we can do are feature flags (in some way, e.g. -preview or now this flag).

RazvanN7 commented:
> Actually implement generating separate header files (the directory passed to -HDd is currently treated as a prefix for the file passed to -HDf

This is not possible (should be read as hard to implement) for the dmd codebase, due to the differences between C++ and D when it comes to importing (vs includes) and forward references.

@edi33416 tried this but it didn't work for the dmd codebase (I can't remember the reason though, it was something about having duplicate forward declarations). You can still generate individual header files, it's just that you have to invoke the compiler for every .d file separately.

MoonlightSentinel commented:
> @Geod24 @MoonlightSentinel Semantic is run before generating the cpp header files. The `generateCPP` function is called in `mars.d` after full semantic analysis has been performed. I think that the reason for this is that you may have structs that have fields that are template instances (`T!int a`). Also, you want to make sure that the code is correct before generating C++ header files.

That is (sadly) not entirely true as some checks are delayed until IR / code generation (and hence occur after the header generation). E.g.
```
extern(C++) void foo(string s) {}
```
fails with
```
Error: Internal Compiler Error: type string cannot be mapped to C++
```
but does not issue an error when using `-c -o-`. Then `-HC` still generates a header containing
```
extern _d_void foo(DArray<_d_char> s);
```

MoonlightSentinel commented:
> > Maybe such big features should be developed on a dedicated branch in the future. That would offer similar benefits to a draft PR while also keeping the work publicly accessible (everbody can make a small PR without adopting the entire stalled PR)
> 
> FYI: We have tried this before and even made nightly builds of these branches available.

Thanks, I did not know that. Seems like it's usually not the best idea.

MoonlightSentinel commented:
> This is not possible (should be read as hard to implement) for the dmd codebase, due to the differences between C++ and D when it comes to importing (vs includes) and forward references.

Agreed, this will be a bigger project for the future.

MoonlightSentinel commented:
What's up with the weird spacing and random braces in the [generated changelog](http://dtest.dlang.io/artifact/website-e5eff32894a30296a32674251ac1f6cf2a0f54a4-757f9606a73ec82db422f08d50fd68ba/web/changelog/pending.html)?

MoonlightSentinel committed:
Move extra header files into TEST_OUTPUT

MoonlightSentinel committed:
Apply further cleanup

- remove unused variables
- sort import in mars.d
- remove unecessary braces
- pointer style
- replace duplication with goto
- TODO comments for missing implementation
- Replace Type[...].ty check + cast with isType[...]

MoonlightSentinel commented:
Eliminated the remaining style issues and got the tests passing by switching from separate files to `TEST_OUTPUT` sections (hence it currently includes #10807 to avoid duplication for `-m32` and `-m32mscoff`).

wilzbach commented:
> What's up with the weird spacing and random braces in the [generated changelog](http://dtest.dlang.io/artifact/website-e5eff32894a30296a32674251ac1f6cf2a0f54a4-757f9606a73ec82db422f08d50fd68ba/web/changelog/pending.html)?

Fix: https://github.com/dlang/tools/pull/390

wilzbach commented on src/dmd/dtoh.d:
```diff
@@ -0,0 +1,1692 @@
+
+private bool isIgnoredModule(ASTCodegen.Module m)
+{
+    //if (!m || !m.parent)
```
relict?

MoonlightSentinel replied:
Probably

wilzbach commented on src/dmd/dtoh.d:
```diff
@@ -0,0 +1,1692 @@
+    {
+        DMDModule._init();
+        DMDClass._init();
+    }
```
You might want to protect these initialization with a singleton, s.t. double initialization doesn't happen when DMD is e.g. used as a library.

MoonlightSentinel replied:
Done

wilzbach commented:
Maybe as a last point to finalize this PR we could add a `runnable`, but there's now even `dshell`.
So could similar to the integration for the hand-written header files, do something like this where we generate the header from a D library and then check whether with the header file app successfully be built and run:

```d
import dshell
void main()
{
     if (OS != "linux")
     {
            writeln("CPP header generation test was skipped on non-linux platform.");
            return;
    }
    Vars.set("header_gen_dir", "$OUTPUT_BASE/cpp_header_gen");
    Vars.set("lib", "$header_gen_dir/a$LIBEXT");
    run("$DMD -m$MODEL -c -of$lib -lib <header-flag> $EXTRA_FILES/header_gen_a.d");
    run("g++ -c $EXTRA_FILES/header_gen_b.cpp <header>"); // cpp-program calls structs, classes etc. from D
   run("$DMD -of$header_gen_dir/test $lib $header_gen_lib/b.o");
   run("$header_gen_dir/test");
}
```

MoonlightSentinel committed:
Address wilzbach's review

- Remove leftover comment
- Guard against multiple initializations
- Add a dshell test

MoonlightSentinel commented:
Added a basic test which covers some basics (and also highlights some existing issues). 
Not sure why it cannot find `libphobos2` on the CI:
```
[RUN] '/home/braddr/sandbox/at-client/pull-3946897-Linux_64_32/dmd/generated/linux/release/64/dmd' '-of=generated/dshell/cpp_header_gen/test' 'generated/dshell/cpp_header_gen/lib.a' 'generated/dshell/cpp_header_gen/cpp.o'
/usr/bin/ld: cannot find -lphobos2
```

Geod24 commented:
> Then `-HC` still generates a header containing [...]

This is probably a DMD-specific hack. Otherwise anything including slices wouldn't be convertible (even structs that are only passed by ref).
It's definitely something that needs to be fixed, see https://github.com/dlang/dmd/pull/8120

@MoonlightSentinel : Please don't use handles in commit messages, otherwise everytime someone pushes the commit we get a notification. Amend, rebase, and keep a clean commit history :)

MoonlightSentinel commented:
> @MoonlightSentinel : Please don't use handles in commit messages, otherwise everytime someone pushes the commit we get a notification. Amend, rebase, and keep a clean commit history :)

Sorry, I forgot that GitHub sends out a notification whenever someone is mentioned in a commit. Your inbox will be safe in the future.

thewilsonator commented on src/dmd/dtoh.d:
```diff
@@ -0,0 +1,1703 @@
+    OutBuffer *fwdbuf;
+    OutBuffer *checkbuf;
+    OutBuffer *donebuf;
+    OutBuffer *buf;
```
please swap the `*` and space.

thewilsonator commented on src/dmd/dtoh.d:
```diff
@@ -0,0 +1,1703 @@
+                    if (!AST.isType(o))
+                        return;
+                }
+                buf.writestring(ti.tempdecl.ident.toChars());
```
this is a duplicate of below. Jut use a `goto` here.

thewilsonator commented on src/dmd/dtoh.d:
```diff
@@ -0,0 +1,1703 @@
+            printf("[AST.RealExp enter] %s\n", e.toChars());
+            scope(exit) printf("[AST.RealExp exit] %s\n", e.toChars());
+        }
+        buf.writestring("0");
```
This looks a bit suspicious, its OK if you want to defer this to a later PR though.

MoonlightSentinel replied:
Thats definitely wrong:

```
extern(C++) struct S
{
    double member = 3.141;
}
```
```
struct S
{
    _d_double member;
    S() : member(0) {}
};
```

But I would prefer to keep the initial version close to the original PRs functionality.

thewilsonator replied:
Ok. Please put a large TODO next then.

MoonlightSentinel replied:
Done

thewilsonator commented on src/dmd/dtoh.d:
```diff
@@ -0,0 +1,1699 @@
+            buf.writestring(";\n");
+            if (vd.type.ty == AST.Tstruct)
+            {
+                auto t = cast(AST.TypeStruct)vd.type;
```
```
if (auto t = vd.type.isStructType())
```
(or whatever isStructType is actually called.

thewilsonator commented on src/dmd/dtoh.d:
```diff
@@ -0,0 +1,1699 @@
+                buf.printf("\n");
+            return;
+        }
+        if (auto fd = ad.aliassym.isDtorDeclaration())
```
`fd` is not used, no need to assign or declare it.

thewilsonator commented on src/dmd/dtoh.d:
```diff
@@ -0,0 +1,1699 @@
+        this.ident = null;
+        if (t.ty == AST.Tsarray)
+        {
+            auto tsa = cast(AST.TypeSArray)t;
```
```
if (auto tsa = t.isStaticArray())
```
(or whatever its called)

thewilsonator reviewed (approved):
Otherwise looks good to me.

RazvanN7 commented:
So what's the plan to remove all the .h files from dmd-fe? We won't be able to market this feature if we aren't using it in-house.

thewilsonator commented:
Step 0: merge this.
Step 1: let @TurkeyMan submit bug reports
Step 2: Fix them.
Step 3: Wait for @ibuclaw to get this into a GCC release.
Step 4: Wait again for that release to be the version used to build subsequent versions of GCC.
Step 5: Make the release that contains fixes from step 2 the oldest version used to build DMD.
Step 6: `rm -rf *.h`

RazvanN7 commented:
Let's get this in! It seems that the buildkite failure is due to some version mismatch. probably unrelated to this PR.

Geod24 commented:
> Let's get this in! It seems that the buildkite failure is due to some version mismatch. probably unrelated to this PR.

Correct: https://github.com/dlang/ci/pull/410

Geod24 commented:
I'll try to give this a review this weekend.

wilzbach commented:
> Step 3: Wait for @ibuclaw to get this into a GCC release.
> Step 4: Wait again for that release to be the version used to build subsequent versions of GCC.
> Step 5: Make the release that contains fixes from step 2 the oldest version used to build DMD.
> Step 6: rm -rf *.h

All these steps aren't required. We talked about this before. It's no problem because:
A) we can check-in the header file into source code (with a CI check enforcing this)
B) DMD frontend requires a working D compiler, so it's absolutely reasonable to except that the build.d build script will be able to generate the headers
I think we still want to go with option A for now as it will provide LDC/GDC with a nice git history of all header file changes for a release.

wilzbach reviewed (approved):
Awesome! Glad to get this finally started!

TurkeyMan commented:
🎉🎉🎉

Awesome! Thank you so much everyone for working this through!

ibuclaw commented:
> A) we can check-in the header file into source code (with a CI check enforcing this)

I think this was my suggestion, it reminds contributors to regenerate the headers (it can be a make rule) before submitting the PR.  Keeping it checked in is harmless anyway.

There will always need to be a header present for the initial bootstrap (assume that for the next 2-3 years, the baseline D compiler doesn't have `-HC`), otherwise you won't be able to build the C++ parts of gdc and ldc..

ibuclaw commented:
Also, this needs a little extra work, as the module in its current form both imports dmd-specific modules and doesn't export the main entrypoint to C++.

Geod24 commented:
Templates seems completely unhandled, and due to the way the FE treats `alias`, I doubt they will be simple to handle.
Also while I was playing around with this I accidentally hit an ICE: https://issues.dlang.org/show_bug.cgi?id=20604

yebblies commented:
From 2015-09-16 to 2020-02-22! (#5082 #8591 #9971 #10801)  It lives!

ibuclaw commented:
@yebblies glad to hear from you as always. :-)
